src/lib/io/VDB.cpp

```diff
@@ -0,0 +1,371 @@
+
+#include "../Partio.h"
+#include "../core/ParticleHeaders.h"
+#include "PartioEndian.h"
+
+#include <openvdb/openvdb.h>
+#include <openvdb/points/PointConversion.h>
+#include <openvdb/points/PointCount.h>
+
+#include <cfloat>
+#include <unordered_map>
+#include <tbb/task_scheduler_init.h>
+namespace Partio
+{
+
+static void initLibrary() {
+    static struct OpenVDBLib {
+        OpenVDBLib()  { openvdb::initialize(); }
+        ~OpenVDBLib() { openvdb::uninitialize(); }
+    } vdblib_init;
+}
+
+template <typename PartioType, typename VDBType, typename IndexType> void
+convertScalar(const openvdb::points::AttributeHandle<VDBType>& vdbAttr,
+              const IndexType&                                 vdbIndex,
+              ParticlesDataMutable*                            particles,
+              ParticleAttribute&                               partioAttr,
+              size_t                                           partioIndex)
+{
+    auto* attrStorage = particles->dataWrite<PartioType>(partioAttr, partioIndex);
+    for (int i = 0; i < partioAttr.count; ++i)
+        attrStorage[i] = vdbAttr.get(vdbIndex, i);
+}
+
+template <typename PartioType, typename VDBType, typename IndexType> PartioType*
+convertVector(const openvdb::points::AttributeHandle<VDBType>& vdbAttr,
+              const IndexType&                                 vdbIndex,
+              ParticlesDataMutable*                            particles,
+              ParticleAttribute&                               partioAttr,
+              size_t                                           partioIndex)
+{
+    openvdb::Vec3d vdbData = vdbAttr.get(vdbIndex);
+    PartioType* partioData = particles->dataWrite<PartioType>(partioAttr, partioIndex);
+
+    for (unsigned i = 0; i < 3; ++i)
+        partioData[i] = vdbData[i];
+
+    return partioData;
+}
+
+enum {
+    Precision32,
+    Precision64
+};
+
+ParticlesDataMutable* readVDB(const char* filename,
+                              const bool headersOnly,
+                              std::ostream* errorStream)
+{
+
+    std::size_t numThreads = 0;
+    char *cueThreads = getenv("CUE_THREADS");
+    if (cueThreads)
+    {
+        numThreads = atoi(cueThreads);
+    }
+    tbb::task_scheduler_init schedulerInit((numThreads == 0) ? tbb::task_scheduler_init::automatic : numThreads);
+
+    initLibrary();
```

Comment:
initLibrary() and all this TBB scheduler initialization stuff seems like it's `main()`'s responsibility rather than the responsibility of a plugin.

`CUE_THREADS` is rather specific and seeing `openvb::uninitialize()` makes me a bit nervous because I'm assuming that it's touching some global state.

I think this would be better mentioned in the readme with some example code showing what users need to do in order to use the library rather than taking on these responsibilities inside library code.

Try to keep these parts generic and reusable with as little assumptions as possible.